Initial edits for GPU page

vyasr opened the pull request:
This PR updates the GPU section of the documentation to capture the current state of NVIDIA/CUDA packaging on PyPI.

vyasr commented:
CC @leofang

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
-PyTorch and TensorFlow). For most of its own projects, it uses a
-[Private PyPI Index](https://pypi.org/project/nvidia-pyindex/) - and that also
-includes rebuilds of TensorFlow and other packages.
+If wheels contain CUDA code, they require the CUDA Toolkit (a specific version
+of it at that) to be installed. Again, installers do not know this and there is
```
Not sure what "a specific version of it _at that_" refers to?

rgommers replied:
It simply means there are version constraints, you must know if it's CUDA `11`, `12`, or even a more specific minor version like `11.1`. Maybe _"they require a specific version of the CUDA Toolkit"_ would be clearer?

vyasr replied:
Perhaps there's just a lack of familiarity with this phrasing at play here? Does [this ref](https://idioms.thefreedictionary.com/at+that) help? In any case, happy to go with Ralf's rewording.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+NVIDIA has made steps towards better support for CUDA on PyPI. Various
+library components of the CTK have been packaged as wheels and are now
+distributed on PyPI, such as
+[nvidia-cublas-cu11](https://pypi.org/project/nvidia-cublas-cu11/). Python
+wrappers around CUDA libraries have been consolidated into CUDA Python
```
Thinking it twice, I feel we may need to phrase it differently. CUDA Python only offers wrappers for driver/runtime/nvrtc, and not CUDA libraries (cuBLAS & co) yet. Maybe changing CUDA "libraries" to CUDA "APIs"?

vyasr replied:
I added that as well as an additional parenthetical afterwards to indicate that there are no bindings for the libraries as of yet.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+wrappers around CUDA libraries have been consolidated into CUDA Python
+([website](https://developer.nvidia.com/cuda-python), 
+[PyPI package](https://pypi.org/project/cuda-python)), but this package assumes
+that the CUDA runtime and driver are already installed since it only provides
+Python bindings to the API. Many other projects remain hosted on
```
nit:
```suggestion
that the CUDA driver and NVRTC are already installed since it only provides
Python bindings to the APIs. Many other projects remain hosted on
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+no way to express this dependency. The same will be true for ROCm and other
+types of GPU hardware and languages.
+
+NVIDIA has made steps towards better support for CUDA on PyPI. Various
+library components of the CTK have been packaged as wheels and are now
+distributed on PyPI, such as
+[nvidia-cublas-cu11](https://pypi.org/project/nvidia-cublas-cu11/). Python
```
btw for the wheel mention, we could add one sentence hyperlinking to the section (not on this page) "Lack of support for symlinks in wheels", such as the following
```
, such as [nvidia-cublas-cu11](https://pypi.org/project/nvidia-cublas-cu11/), although special care is needed to consume them due to [lack of symlinks in wheels](...).
```

vyasr replied:
Done.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
 [`cupy-cuda111`](https://pypi.org/project/cupy-cuda111/),
-[`cupy-cuda11x`](https://pypi.org/project/cupy-cuda11x/),
 [`cupy-rocm-4-3`](https://pypi.org/project/cupy-rocm-4-3/),
 [`cupy-rocm-5-0`](https://pypi.org/project/cupy-rocm-5-0/).
```
I am thinking to add a sentence or two here explaining why it's painful to everyone.
```suggestion
[`cupy-rocm-5-0`](https://pypi.org/project/cupy-rocm-5-0/). This not only adds maintenance
overhead to project developers and consumes more storage and network bandwidth for PyPI.org,
but also prevents downstream projects from properly declaring the dependency.
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
-`manylinux2014` wheels for TensorFlow are 588 MB
+As of CUDA 11, CUDA promises [minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility),
+which allows building packages compatibility across an entire CUDA major
```
```suggestion
which allows building packages compatible across an entire CUDA major
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+[`cupy-cuda12x`](https://pypi.org/project/cupy-cuda12x/) that work for any CUDA
+11.x or CUDA 12.x version, respectively, that a user has installed.
+
+GPU packages tend to result in very large wheels. This is in large part because
```
nit
```suggestion
GPU packages tend to result in very large wheels. This is mainly because
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+GPU packages tend to result in very large wheels. This is in large part because
+compiled GPU libraries must support a number of architectures, leading to large
+binary sizes. These effects are compounded by the requirements imposed by the
+manylinux standard for Linux wheels, which results in many large libraries being
```
I wonder if we should add a link to the "Depending on packages for which an ABI matters" section (on another page) since we're touching manylinux here. No strong opinion, though.

vyasr replied:
I'm not opposed, but I'm not sure exactly how that would look here. Perhaps just a "see also [the notes on ABI compatibility](...)"?

rgommers replied:
I think it's actually more this page where vendoring requirements are explained: https://pypackaging-native.github.io/key-issues/native-dependencies/. So I suggest adding:
```
(see [Native dependencies](native-dependencies) for details)
```

leofang reviewed (commented):
Thanks Vyas! Added a few comments. 

Could you add a reference to [our demo](https://github.com/NVIDIA/cuQuantum/tree/main/extra/demo_build_with_wheels) to the "Relevant resources" section at the end of this page? Originally I was thinking to migrate text from there, but your revision fits the original flow better (less changes while correctly covering all the information needed), so perhaps a cross reference is sufficient.

rgommers commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
-PyTorch and TensorFlow). For most of its own projects, it uses a
-[Private PyPI Index](https://pypi.org/project/nvidia-pyindex/) - and that also
-includes rebuilds of TensorFlow and other packages.
+If wheels contain CUDA code, they require the CUDA Toolkit (a specific version
```
Can you add (CTK) as the abbreviation here? It's used further down like that.

rgommers commented on docs/key-issues/gpus.md:
```diff
@@ -30,52 +30,66 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
-for deep learning packages, because they link in
-[cuDNN](https://developer.nvidia.com/cudnn). For example, the most recent
-`manylinux2014` wheels for TensorFlow are 588 MB
+As of CUDA 11, CUDA promises [minor version
```
IIRC the first version to actually achieve this was CUDA 11.2. Is that right? If so, it may be worth adding ".2"

leofang replied:
Technically it starts with 11.0, with the exception of NVRTC (which CuPy et al depends on) until 11.2. Also, minor ver compat in CUDA 11 does not cover PTX compilation. Not sure if theses technical details are important here. No strong opinion.

rgommers replied:
Maybe as a footnote? Seems like the kind of thing that the majority of folks aren't interested in, but if you need it it's quite important and not easy to find.

vyasr replied:
Hmm yeah this is a bit tricky. How deep into the weeds do we want to get here? If we mention PTX compilation, should we also mention that this is specifically an issue for `numba.cuda`, which provides additional plugins to get around this (the ptxcompiler and cubinlinker packages)? Do we want to mention the constraints for any other package that distributes PTX code (they need to restrict the PTX to instructions compiled for the oldest driver they want to use, which in practice means they probably need to compile on the oldest runtime they want to use, within a given major version)? What about how that applies to JIT outside of numba (for instance cupy has the ability to JIT, although I think in the long run there is interest in consolidating with numba, Leo could say more here)?

vyasr replied:
Also the differences between CUDA 10 (which I imagine many stacks are still using somewhere), CUDA 11, and CUDA 12?

rgommers replied:
If we could add the compat details and the PTX issue, with perhaps a link to a relevant Numba issue, in a footnote that would be really nice. It's the kind of info that is very difficult to find and hence valuable to have available.

CUDA 10 I'm less sure about, I'd be fine with leaving it out on the basis everyone is moving or has moved away from it.

rgommers commented:
Thanks for doing this @vyasr!

vyasr commented:
Happy to help! 

> Could you add a reference to [our demo](https://github.com/NVIDIA/cuQuantum/tree/main/extra/demo_build_with_wheels) to the "Relevant resources" section at the end of this page? Originally I was thinking to migrate text from there, but your revision fits the original flow better (less changes while correctly covering all the information needed), so perhaps a cross reference is sufficient.

Done. In the future if RAPIDS gets around to writing up our process we could include that as well.

rgommers commented:
> In the future if RAPIDS gets around to writing up our process we could include that as well.

Yes please, that'd be great to link to.

rgommers commented on docs/key-issues/gpus.md:
```diff
@@ -201,3 +204,84 @@ Potential solutions on the PyPI side include:
+  access with CUDA C(++). However, since PTX code does not support MVC, PTX
+  compiled with a particular CUDA runtime may not work if run on a system with
+  an older driver. This fact has two consequences. First, libraries that
+  package CUDA code will not benefit from MVC. Second, libraries that leverage
```
This first consequence assumes that the library is shipped with PTX code, right? If not, I'm not sure I understand that sentence.

rgommers reviewed (commented):
This added content on CUDA compatibility is super useful, thanks @vyasr!

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+it on a system with a CUDA 11.2 driver installed. However there are some
+caveats with MVC:
+
+- If the code uses any features that were introduced in a later driver version
```
Is `cuGetProcAddress`, introduced in CUDA 11.3, an exception to this? If I attempt to call it using `ctypes` in Python or `dlsym` in C, will I just get a normal linker error?

gmarkall replied:
Now I think about it, is this a feature of the runtime, rather than a guarantee made for code that directly uses the driver API?

vyasr replied:
Yeah I was under the impression that MVC only applies to the runtime API, not the driver API. We could test that though to be sure.

leofang replied:
Minor ver compat applies to the UMD too.

gmarkall replied:
Looking over the docs, I can't see anything that specifically talks about direct use of driver APIs - the closest thing I can find is in https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#handling-new-cuda-features-and-driver-apis, which says:

> To use other CUDA APIs introduced in a minor release (that require a new driver), one would have to implement fallbacks or fail gracefully.

@leofang Have you seen anything that refers more specifically to driver APIs?

For my example of `cuGetProcAddress`, I've seen it suggested (elsewhere) that one should fall back to a platform-specific equivalent - e.g. `GetProcAddress()` on Windows, or `dlsym()` on Linux. Maybe we need to note here that direct use of driver APIs added in minor releases needs to be considered / guarded against appropriately, without being too prescriptive (since the thing to do could vary depending on what the API is and the application features / requirements).

leofang replied:
Yes, we need fallbacks for new APIs introduced in the middle of a major version. But this is nothing unusual. MVC is basically reaffirming that CUDA is obeying semantic versioning IMHO, and it's OK to make non-breaking changes (ex: add new APIs) in a minor release, so I am not sure we should expand on this.

vyasr replied:
If a new API is introduced in CUDA 12.2, then any code using it will need fallbacks/failure modes handling it when compiling on CUDA 12.0/12.1. That's true whether it's a runtime or driver API, which I think is also what Leo is saying.

I'm not really sure how minor version compatibility makes sense for the UMD. The driver APIs are implemented in the same library that provides the UMD (libcuda.so). How would you have a "newer" driver API installed than the UMD?

The whole point of MVC IIUC is that using code compiled against a specific libcudart won't automatically give linker errors if that libcudart finds an older libcuda on the system than the one it was compiled against.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  solution to this problem is for libraries to use runtime checks of the CUDA
+  version (using e.g. `cudaDriverGetVersion`) to only use supported features on
+  the installed driver.
+- NVRTC did not start supporting MVC until CUDA 11.3. Therefore, code that uses
```
Additionally, I believe that you cannot use NVRTC to generate code that needs linking by the driver under MVC (e.g. when calling device functions from another translation unit) - it will produce a cubin, but the driver will refuse to add it to a link (with a cryptic error code, IIRC).

vyasr replied:
Are there specific version constraints here? Is this true for any 11.x NVRTC?

gmarkall replied:
I think it's true for any NVRTC that supports MVC, including in CUDA 12 - you would still need nvJitlink for MVC with linking.

leofang replied:
This is my understanding:
- NVRTC -> compiler (CUDA C++ -> PTX or SASS)
- nvJitLink -> linker

and to support MVC we need both pieces, unless linking is not needed (ex: compiling a single TU) as Graham noted. 

That nvJitLink was missing in CUDA 11 is one reason that MVC could not cover a lot of PTX use cases. Another reason is that the UMD driver is responsible for compiling (JIT'ing) PTX to SASS, but if the PTX is generated with NVRTC newer than UMD (say, NVRTC from 11.6, UMD from 11.2), JIT'ing will fail, too. (Reason: Each CTK minor release bumps the PTX minor version -- yes, it has its own version -- so an old driver would simply not recognize the new code.) This was why CuPy had to compile down to SASS with NVRTC.

vyasr replied:
Hmm @leofang [this diagram](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#deployment-considerations) indicates 11.3, not 11.2. Is it wrong?

Aside from that this discussion makes sense though, will update.

vyasr replied:
The single TU + no linking restriction is now mentioned below.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  the installed driver.
+- NVRTC did not start supporting MVC until CUDA 11.3. Therefore, code that uses
+  NVRTC for JIT compilation must have been compiled with a CUDA version >= 11.3.
+- MVC only applies to CUDA code, not
```
I think this is the specific code you mean here?

```suggestion
- MVC only applies to SASS binary code, not
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  instruction set that the CUDA driver library can JIT-compile to machine
+  instructions. The standard [CUDA compilation
+  pipeline](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#the-cuda-compilation-trajectory)
+  includes the translation of CUDA code into PTX, but in addition developers
```
"code" alone is a little vague:

```suggestion
  includes the translation of CUDA source code into PTX, but in addition developers
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  instructions. The standard [CUDA compilation
+  pipeline](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#the-cuda-compilation-trajectory)
+  includes the translation of CUDA code into PTX, but in addition developers
+  may write PTX code directly rather than CUDA for performance since it offers
```
I think suggesting the two use cases upfront makes the rest of the paragraph easier to digest:

```suggestion
  may write PTX code directly for performance, or use code generation tools that produce PTX, since it offers
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  includes the translation of CUDA code into PTX, but in addition developers
+  may write PTX code directly rather than CUDA for performance since it offers
+  avenues for optimization that are either unavailable or very difficult to
+  access with CUDA C(++). However, since PTX code does not support MVC, PTX
```
I think this is a more common idiom - I've not come across `(++)` usage (much?) before

```suggestion
  access with CUDA C/C++. However, since PTX code does not support MVC, PTX
```

gmarkall replied:
Also, I think the supporter / supported feels inverted here:

```suggestion
  access with CUDA C(++). However, since MVC does not support PTX code, PTX
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  may write PTX code directly rather than CUDA for performance since it offers
+  avenues for optimization that are either unavailable or very difficult to
+  access with CUDA C(++). However, since PTX code does not support MVC, PTX
+  compiled with a particular CUDA runtime may not work if run on a system with
```
As I think of it, NVCC and NVRTC (via NVVM) are toolkit components that generate the PTX - I think suggesting the "runtime" generates it feels confusing.

```suggestion
  generated using a particular CUDA toolkit may not work if run on a system with
```

vyasr replied:
Good call, I agree I'm conflating the two a bit here.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  package PTX code will not benefit from MVC. Second, libraries that leverage
+  any sort of JIT-compilation pipeline that generates PTX code will _also_ not
+  support MVC. The latter can lead to more surprising behaviors, such as if a
+  user has a newer CUDA runtime than driver and then uses
```
Another use of "runtime", where I think the toolkit version is really what's important:

```suggestion
  user has a newer CUDA toolkit than driver and then uses
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  support MVC. The latter can lead to more surprising behaviors, such as if a
+  user has a newer CUDA runtime than driver and then uses
+  [numba.cuda](https://numba.pydata.org/) to compile a Python function since
+  numba compiles CUDA kernels to PTX as part of its pipeline. Prior to CUDA
+  12, CUDA itself provides no solutions to this problem, although in some cases
```
I don't think this is strictly true - since 11.1 or 11.2 (I can't recall exactly when) `libnvptxcompiler_static.a` has been bundled with the toolkit in order to support JIT compilation of PTX for older drivers than the toolkit used for the build - the build links in `libnvptxcompiler_static.a` and uses it to compile PTX bundled or generated at runtime. [PTX Compiler APIs](https://docs.nvidia.com/cuda/ptx-compiler-api/index.html) documents the usage of the library and API. The addition RAPIDS made on top of this was to build a Pythonic binding capable of patching Numba to use its bundled version of libnvptxcompiler_static which we build with 11.8, so it can always compile any 11.x PTX (because 11.8 was the last 11.x release).

This is only a partial solution - if multiple PTX files need to be linked, then there isn't an official general solution, but RAPIDS also provides the cubinlinker package, which provides a Pythonic binding to some internal APIs to address both PTX compilation and linking.

vyasr replied:
True, there are some partial solutions available. IMO the distinction between compiling a single PTX file and linking multiple is too deep in the weeds even for this document. I personally wouldn't even know how to describe exactly when the former is sufficient and when the latter would be necessary. I'd love to fill that gap in my own knowledge, but I think adding that here at this point would be a bit too overwhelming for readers who aren't familiar with the entire CUDA compilation pipeline.

Do you think there's a reasonable middle ground in what can be stated here? In particular, if we focus on it from the perspective of a developer trying to make their package run: for what use cases are the PTX compiler APIs sufficient, and when would something like cubinlinker be required?

gmarkall replied:
> Is it accurate to state that this basically means that PTX functions cannot call PTX functions from another TU?

It's accurate, but it also doesn't elaborate on the implications. That TU might be from another PTX file provided by the user, or it might be in a library distributed as part of the toolkit, e.g. `libcudadevrt.a`, which contains functions used e.g. for dynamic parallellism or cooperative groups. So it also means that certain CUDA functionality becomes unavailable.

gmarkall replied:
> Can CUDA source code from another TU call PTX functions compiled using the PTX compiler API, or is the limitation bidirectional?

The limitation is bidirectional, because you can't add anything compiled for MVC to a link.

gmarkall replied:
> I assume that the latter works, otherwise it's fairly useless to have a completely isolated compilation unit that is JIT-compiled but can't be used by anything else.

The majority of Numba functionality can be supported in this way, it's not completely useless :slightly_smiling_face: - just not cooperative groups and linking C/C++ device code.

> I'd love to fill that gap in my own knowledge, but I think adding that here at this point would be a bit too overwhelming for readers who aren't familiar with the entire CUDA compilation pipeline.

Indeed, it is very complicated to spell out exactly what is and isn't possible, especially in generic terms and not "which subset of Numba can work in this case".

> Do you think there's a reasonable middle ground in what can be stated here? In particular, if we focus on it from the perspective of a developer trying to make their package run: for what use cases are the PTX compiler APIs sufficient, and when would something like cubinlinker be required?

I think maybe the middle ground just looks like this:

> CUDA itself provides no **general** solutions to this problem

(addition of the word "general") and leaving it at that, otherwise we end up discussing all these nuances and losing the wood for trees.

vyasr replied:
OK sounds good. Done.

vyasr replied:
I'm confused at how numba works given the bidirectionality. If a numba-compiled kernel can't link to anything or be linked to, how does anything call it?

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
+  0.57](https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023).
+  CUDA 12 introduces the
+  [nvJitLink](https://docs.nvidia.com/cuda/nvjitlink/index.html) library as the
+  long-term solution to this problem. nvJitLink may be leveraged to compile PTX
```
It does, but the Python binding is still WIP (https://github.com/gmarkall/nvjitlink) and Numba and any other applications will need modification to use it.

vyasr replied:
I think we can probably leave that link out for now and simply state that Python bindings do not exist yet, then update the page when they do. Otherwise we'll be sending readers down a potentially unhelpful rabbit hole, especially if the bindings move into a different org or library over time.

leofang replied:
nvJitLink is the ultimate solution, but with its own caveats. To link together different pieces generated from different minor CTK versions, it requires to use nvJitLink coming from a CTK whose minor version is higher than those of all participating pieces (see the [CUDA minor version compatibility](https://developer.nvidia.com/blog/cuda-12-0-compiler-support-for-runtime-lto-using-nvjitlink-library/) section; need to search in the article, as there's no direct hyperlink, unfortunately). Ex: if I'm linking pieces generated from CTK 12.0, 12.1, and 12.2, I need to use nvJitLink from 12.2+.

gmarkall replied:
> I think we can probably leave that link out for now and simply state that Python bindings do not exist yet, then update the page when they do. Otherwise we'll be sending readers down a potentially unhelpful rabbit hole, especially if the bindings move into a different org or library over time.

Indeed, I didn't intend to imply we should add that link - just pointing to where the current state of things is :-)

vyasr commented:
Thanks for the review @gmarkall! I've addressed most of your smaller points and left follow-up questions for the others.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  instruction set that the CUDA driver library can JIT-compile to machine
+  instructions. The standard [CUDA compilation
+  pipeline](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#the-cuda-compilation-trajectory)
+  includes the translation of CUDA source code into PTX. In addition, to
+  achieve the best performance, developers may write PTX code directly or use
+  code generation tools that produce PTX because it offers avenues for
+  optimization that are either unavailable or very difficult to access with
+  CUDA C/C++.  However, since MVC does not support PTX code, PTX generated
```
While this is true, I think it's really too detailed. We simply wanna mention that some projects choose to generate PTX here, without justifying for them. I suggest the following change:
```suggestion
  includes the translation of CUDA source code into PTX. In addition, some projects
  choose to include PTX code generated either at build time or run time.
  However, since MVC does not cover JIT'ing PTX code, PTX generated
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  code generation tools that produce PTX because it offers avenues for
+  optimization that are either unavailable or very difficult to access with
+  CUDA C/C++.  However, since MVC does not support PTX code, PTX generated
+  using a particular CUDA toolkit may not work if run on a system with an older
+  driver.  This fact has two consequences. First, libraries that package PTX
```
```suggestion
  using a particular CTK may not work with an older driver. 
  This fact has two consequences. First, libraries that package PTX
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  code will not benefit from MVC. Second, libraries that leverage any sort of
+  JIT-compilation pipeline that generates PTX code will _also_ not support MVC.
+  The latter can lead to more surprising behaviors, such as if a user has a
+  newer CUDA toolkit than driver and then uses
```
```suggestion
  newer CTK than the installed driver and then uses
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  driver.  This fact has two consequences. First, libraries that package PTX
+  code will not benefit from MVC. Second, libraries that leverage any sort of
+  JIT-compilation pipeline that generates PTX code will _also_ not support MVC.
+  The latter can lead to more surprising behaviors, such as if a user has a
```
I think we meant "when" not "if" here, because we didn't mention how the behavior is surprising?
```suggestion
  The latter can lead to more surprising behaviors, such as when a user has a
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  [numba.cuda](https://numba.pydata.org/) to compile a Python function since
+  numba compiles CUDA kernels to PTX as part of its pipeline. Prior to CUDA 12,
+  CUDA itself provides no solutions to this problem, although in some cases
+  there are tools that may be used to help (for instance, numba supports MVC in
```
```suggestion
  there are tools that may help (for instance, Numba supports MVC in
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  JIT-compilation pipeline that generates PTX code will _also_ not support MVC.
+  The latter can lead to more surprising behaviors, such as if a user has a
+  newer CUDA toolkit than driver and then uses
+  [numba.cuda](https://numba.pydata.org/) to compile a Python function since
+  numba compiles CUDA kernels to PTX as part of its pipeline. Prior to CUDA 12,
```
Is this really true? @gmarkall I thought Numba usually bypasses PTX generation, unless compiling CUDA C++ headers or linking to other TUs that provide device functions is needed?

gmarkall replied:
Yes, it's true - Numba's pipeline goes:

Python source -> Python bytecode -> Numba IR -> LLVM (NVVM) IR -> PTX -> SASS.

When CUDA C++ headers are compiled, they are compiled to PTX then SASS, then linked.

I think a common misunderstanding I've tried to correct before is that people thought Numba translated Python source into CUDA C++ source (like how Warp or CuPy's Python JIT do, IIRC) - could this be what you're thinking of?

leofang commented:
FYI, I learned from Cliff W this great summary, rephrasing in my own words
- CUDA Forward compat: Support newer UMD with older KMD
- CUDA Minor ver compat: Support newer CTK with older UMD

So, if we do the combo right (meeting each compat's requirement), it means we can support newer CTK with older KMD, which is the core story for DC/HPC deployment.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+CUDA also promises [minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility)
+(MVC). This compatibility guarantees that CUDA code compiled using a certain
+version of the CUDA toolkit will be compatible with any driver version within
+the same major release. This behavior is useful because it is often easier for
```
What is "the minimal requirement" in this context?

gmarkall replied:
Should this then say something like:

```suggestion
version of the CUDA toolkit will be compatible with any driver version above
the minimal requirement specified for MVC with that toolkit. This behavior is useful because it is often easier for
```

I'm not super happy with my suggestion - it's a bit of a mouthful - but otherwise I think it's a bit confusing what the minimal requirement is or where to look for it.

vyasr replied:
Alternatively we could just link to [the official table](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#default-to-minor-version). Then we don't have to maintain much.

What @leofang said above is indeed a more precise statement, but only because MVC was only imperfectly included in the first few releases of CUDA 11. For future versions I would expect all the above statements to be equivalent because the earliest official release for major version X will have a UMD compatible with all later minor CTKs.

vyasr commented:
Apologies, this week was overtaken by a few extended firefighting exercises. I'll get back to this next week.

vyasr commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
-separate wheelhouse (see
+package name.
+
+Historically, this required projects to produce packages specific CUDA minor
```
Used "to". Thanks, I missed this typo.

vyasr commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
   declare a dependency on a CUDA or ROCm version as externally provided,
+
+
+## Additional Notes on CUDA Compatibility
```
Alternatively I could just drop the `.so` and use `libcudart` and `libcuda` without suffixes, WDYT?

leofang replied:
Even the prefix "lib" is Unix-only (ex: `libcuda.so` is called `nvcuda.dll` 😇), I wouldn't bother...

vyasr replied:
Good point, I wasn't thinking about that but you're absolutely right. I'll leave it as is for now.

vyasr commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+   typically considered a "driver" in common parlance when referring to other
+   peripherals connected to a computer.
+
+CUDA developers and users generally do not need to be aware of the kernel-mode driver.
```
I would argue that CUDA forward compat is only relevant for HPC _admins_, whereas users and developers do not usually care about it. I would like to leave forward compat out of this discussion altogether to avoid muddying the waters.

leofang replied:
Following up here: https://github.com/pypackaging-native/pypackaging-native/pull/36#discussion_r1239276086

vyasr replied:
Resolved on the other thread.

vyasr commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,85 @@ Potential solutions on the PyPI side include:
+  NVRTC for JIT compilation must have been compiled with a CUDA version >= 11.3.
+- MVC only applies to SASS binary code, not
+  [PTX](https://docs.nvidia.com/cuda/parallel-thread-execution/). PTX is an
+  instruction set that the CUDA driver library can JIT-compile to machine
+  instructions. The standard [CUDA compilation
```
Edited the beginning of this bullet instead to clarify what SASS is.

h-vetinari commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
 ## Current state
 
-As of December 2022, PyPI and Python packaging tools are completely unaware of
+As of May 2023, PyPI and Python packaging tools are completely unaware of
```
Months pass quickly (hey, it's June already!). How about the "As of mid 2023"?

vyasr replied:
Good call, updated.

h-vetinari commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+This works, but adds maintenance overhead to project developers and consumes
+more storage and network bandwidth for PyPI.org. Moreover, it also prevents
+downstream projects from properly declaring the dependency unless they also
+follow a similar multi-package approach. As of CUDA 11, CUDA promises
```
```suggestion
follow a similar multi-package approach. As of CUDA 11.2, CUDA promises
```
If I'm not mistaken, this only started with 11.2

leofang replied:
Depend on which component you're referring to. For NVRTC, yes.

h-vetinari commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+bundled into a single wheel (see [Native dependencies](native-dependencies) for
+details). This is true in particular for deep learning packages because they
+link in [cuDNN](https://developer.nvidia.com/cudnn). For example, recent
+`manylinux2014` wheels for TensorFlow are 588 MB ([2.11.0
+files](https://pypi.org/project/tensorflow/2.11.0/#files)), and for PyTorch
+those are 890 MB ([1.13.0
+files](https://pypi.org/project/torch/1.13.0/#files)). The problems around and
```
Can we call the links here "wheels for v2.11.0" and "wheels for v1.13.0"? I find "2.11.0 files" unintuitive.

vyasr replied:
Agreed, changing.

h-vetinari commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+
+CUDA drivers have always been backward compatible. Any code that runs with some
+driver version X is installed will always run correctly with some newer driver
+version Y>X. As briefly discussed above, as of CUDA 11.0 CUDA also promises
```
11.2 again?

h-vetinari commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+version of the CTK will be compatible with any driver version within the same
+major release. This behavior is useful because it is often easier for users to
+upgrade their CUDA runtime than it is to upgrade the driver, especially on
+shared machines. For instance, the CTK may be installed using conda, while the
+driver library cannot be. An example of leveraging MVC would be compiling code
```
This is changing nowadays. NVidia is [working](https://github.com/conda-forge/staged-recipes/issues/21382) to redistribute its entire[^1] setup through conda-forge.

[^1]: AFAIU; I admit that I don't know if the actual driver library will be part of `cuda`, but it would be consistent with distributing pretty much everything else that used to not be redistributable.

vyasr replied:
Yup, both Leo and I are deeply involved in that effort (FYI we are both NVIDIA employees). The driver library will unfortunately still not be part of the conda-forge offering, but the rest of the CTK will be on cf.

h-vetinari reviewed (commented):
Some comments. I feel there are a lot more things to say[^1], but that's not a criticism of the existing PR, which is a big improvement already.

[^1]: e.g. problems with compatibility between nvcc and platform compilers/stdlib, topic tie-ins with OpenMP, threading & "complex C++ deps", how redistribution is being completely overhauled and granularized in conda-forge, the different degrees of CUDA dependency (e.g. just bare `libcuda.so`, or also other libs [still MVC], up to ptx-usage [no MVC]), understanding the parallel `cu*` world (e.g. cuX vs. original X), unobvious relationship between NVIDIA driver version and CUDA, actually achieving device offloading in various compilation scenarios, etc.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+package](https://pypi.org/project/cuda-python)), but this package assumes that
+the CUDA driver and NVRTC are already installed since it only provides Python
+bindings to the APIs (and no bindings for CUDA libraries are provided as of
+yet). Many other projects remain hosted on [NVIDIA's Private PyPI
```
I know the project description page describes this as "the NVIDIA Private Python Package Index", but does the word "private" convey something else? I'd assume that a "private" index would require some sort of explicit authorization or be inaccessible to the general public - would putting

```suggestion
yet). Many other projects remain hosted on [NVIDIA's PyPI
```

instead convey the idea more straightforwardly? (or is "Private PyPI" a generally-understood term I've not come across before?)

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+of TensorFlow and other packages.
 
 A single CUDA version supports a reasonable range of GPU architectures. New
 CUDA versions get released regularly, and - because they come with increased
 performance or new functionality - it may be necessary or desirable to build
```
I know you didn't modify this, but as-is it's hard to parse - how about

```suggestion
CUDA versions get released regularly, and because they come with increased
performance or new functionality it may be necessary or desirable to build
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+more storage and network bandwidth for PyPI.org. Moreover, it also prevents
+downstream projects from properly declaring the dependency unless they also
+follow a similar multi-package approach. As of CUDA 11, CUDA promises
+[minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility),
```
This is the first use of Minor Version Compatibility, so it should define the acronym:

```suggestion
[Minor Version Compatibility
(MVC)](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility),
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+version. CuPy now leverages this to produce wheels like 
+[`cupy-cuda11x`](https://pypi.org/project/cupy-cuda11x/) and
+[`cupy-cuda12x`](https://pypi.org/project/cupy-cuda12x/) that work for any CUDA
+11.x or CUDA 12.x version, respectively, that a user has installed. Libraries
```
It feels like the commas unnecessarily break up the flow:

```suggestion
11.x or CUDA 12.x version respectively that a user has installed. Libraries
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+version. CuPy now leverages this to produce wheels like 
+[`cupy-cuda11x`](https://pypi.org/project/cupy-cuda11x/) and
+[`cupy-cuda12x`](https://pypi.org/project/cupy-cuda12x/) that work for any CUDA
+11.x or CUDA 12.x version, respectively, that a user has installed. Libraries
+that package PTX code cannot take advantage of this process yet, however ([see
+below](#additional-notes-on-cuda-compatibility)).
```
Does this read a little more straightforwardly?

```suggestion
11.x or CUDA 12.x version, respectively, that a user has installed. However,
libraries that package PTX code cannot take advantage of this process yet
([see below](#additional-notes-on-cuda-compatibility)).
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+
+1. The CUDA Toolkit (CTK): This component includes the CUDA runtime library
+   (libcudart.so) along with a range of other libraries and tools including
+   math libraries like cuBLAS. libcudart.so and a few other core headers and
+   libraries compose the bare minimum required to compile CUDA code.
```
Shared objects aren't involved in compilation - is this more accurate (or needlessly pedantic?)?

```suggestion
   math libraries like cuBLAS. libcudart.so and a few other core headers and
   libraries are the bare minimum required to compile and link CUDA code.
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+   math libraries like cuBLAS. libcudart.so and a few other core headers and
+   libraries compose the bare minimum required to compile CUDA code.
+2. The user-mode driver: This is the libcuda.so library. This library is
+   required to actually run CUDA code.
```
In contrast to compiling and linking (from point 1 just above), does this explain more explicitly what it does?

```suggestion
   required to load and run CUDA code.
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+   (libcudart.so) along with a range of other libraries and tools including
+   math libraries like cuBLAS. libcudart.so and a few other core headers and
+   libraries compose the bare minimum required to compile CUDA code.
+2. The user-mode driver: This is the libcuda.so library. This library is
```
"UMD" and "KMD" are acronyms commonly used, so I wonder if it is helpful to define them here to help the user spot their meaning when studying related docs:

```suggestion
2. The User-Mode Driver (UMD): This is the libcuda.so library. This library is
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+   libraries compose the bare minimum required to compile CUDA code.
+2. The user-mode driver: This is the libcuda.so library. This library is
+   required to actually run CUDA code.
+3. The kernel-mode driver: The nvidia.ko file. This constitutes what is
```
```suggestion
3. The Kernel-Mode Driver (KMD): The nvidia.ko file. This constitutes what is
```

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+   typically considered a "driver" in common parlance when referring to other
+   peripherals connected to a computer.
+
+CUDA developers and users usually do not need to be aware of the kernel-mode driver.
+For the rest of this section, when referring to the "driver" we will always be referring to the user-mode driver libcuda.so.
```
Small edits for clarity:

```suggestion
CUDA developers and users usually do not need to be aware of the Kernel-Mode Driver.
For the rest of this section, when referring to the "driver" we will always be referring to the User-Mode Driver, libcuda.so.
```

leofang replied:
If acronyms are already defined at this point, let's just use them as much as we can.

gmarkall commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+  linking is not possible without nvJitLink (see below).
+- MVC only applies to machine instructions (SASS), not
+  [PTX](https://docs.nvidia.com/cuda/parallel-thread-execution/). PTX is an
+  instruction set that the CUDA driver library can JIT-compile to SASS). The
```
Stray closing parenthesis:
```suggestion
  instruction set that the CUDA driver library can JIT-compile to SASS. The
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+distributed on PyPI, such as
+[nvidia-cublas-cu11](https://pypi.org/project/nvidia-cublas-cu11/), although
+special care is needed to consume them due to [lack of symlinks in
+wheels](native-dependencies/cpp_deps.md#current-state). Python wrappers around
```
The hyperlink is used to refer to the fact (lack of symlink), not to what special care is needed

vyasr replied:
Do either of you see a potential change here that would improve clarity?

vyasr replied:
We discussed this offline and concluded that the current wording is good enough for now.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -169,7 +193,7 @@ support in packaging tools, and lack of people to work on a solution.
 ## Relevant resources
 
-TODO
+- [How cuQuantum builds wheels](https://github.com/NVIDIA/cuQuantum/tree/main/extra/demo_build_with_wheels)
```
I think this is more accurate:
```suggestion
- [How to use cuQuantum wheels](https://github.com/NVIDIA/cuQuantum/tree/main/extra/demo_build_with_wheels)
```

vyasr commented:
@rgommers @leofang @h-vetinari @gmarkall thank you all for the reviews! I think I've addressed everything now.

In the mindset of the perfect being the enemy of the good, I think this PR is now enough of an improvement that we should aim to get it merged soon without too many larger changes. Happy to make edits if there are errors that need correcting or small cosmetic changes, but at this point I think we'd be better served putting off more substantive additions to subsequent PRs. This info is already enough of an improvement that I think it's worth having it go live ASAP. We can open other issues (and I'm happy to make follow-up PRs) as needed. Does that sound reasonable?

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+[Minor Version Compatibility
+(MVC)](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility),
+which allows building packages compatible across an entire CUDA major
+version. CuPy now leverages this to produce wheels like 
```
```suggestion
version. For example, CuPy now leverages this to produce wheels like 
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+   typically considered a "driver" in common parlance when referring to other
+   peripherals connected to a computer.
+
+CUDA developers and users usually do not need to be aware of the KMD.
+For the rest of this section, when referring to the "driver" we will always be referring to the UMD.
```
Sorry, I still don't feel comfortable about the claim that "don't need to be aware of KMD." People who own a workstation under their desk (like I did) would likely disagree, because this bites users from time to time. Can we rephrase it this way? 
```suggestion
As far as MVC is concerned, KMD can be ignored providing that it meets [CUDA's requirement](https://docs.nvidia.com/deploy/cuda-compatibility/index.html).
For the rest of this section, therefore, the "driver" and the UMD are considered interchangeable.
```

vyasr replied:
Fair enough, you're right. I've definitely observed >0 issues with that before, so it can't be ignored altogether. I'll adopt your rewriting of the first sentence.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+CUDA developers and users usually do not need to be aware of the KMD.
+For the rest of this section, when referring to the "driver" we will always be referring to the UMD.
+
+The CUDA runtime library makes no forward or backward compatibility
+guarantees, meaning that libraries that dynamically link to the CUDA runtime
```
(Just leaving a comment, no change is needed.) TBH it's surprising that the CUDA docs described it this way. I would think it's not carefully worded. If cudart had not followed the SemVer semantics for API/ABI compatibility across minor versions (which it actually does), there would have been all sorts of issues.

vyasr replied:
Yeah I agree that the wording is surprising and that there's definitely pretty strong compatibility. I think it's more that the CUDA team reserves the right to break this if needed, and historically they may have taken more advantage of that, but nowadays they have too many people relying on ABI compatibility to break it in practice even though they reserve the right to do so in theory.

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+  version than the installed version, it will still fail to run. However, it
+  will be a runtime failure in the form of a `cudaErrorCallRequiresNewerDriver`
+  CUDA error, rather than a linker error or some similarly opaque issue. One
+  solution to this problem is for libraries to use runtime checks of the CUDA
```
```suggestion
  solution to this problem is for libraries to use runtime checks of the driver
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+  no linking because linking is not possible without nvJitLink (see below).
+- MVC only applies to machine instructions (SASS), not
+  [PTX](https://docs.nvidia.com/cuda/parallel-thread-execution/). PTX is an
+  instruction set that the CUDA driver library can JIT-compile to SASS. The
```
```suggestion
  instruction set that the CUDA driver can JIT-compile to SASS. The
```

leofang commented on docs/key-issues/gpus.md:
```diff
@@ -180,3 +204,87 @@ Potential solutions on the PyPI side include:
+  CUDA itself provides no general solutions to this problem, although in some
+  cases there are tools that may help (for instance, Numba supports
+  MVC in CUDA 11 [starting with numba
+  0.57](https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023).
```
missing enclosing )
```suggestion
  0.57](https://numba.readthedocs.io/en/stable/release-notes.html#version-0-57-0-1-may-2023)).
```

vyasr replied:
Good find 😄

leofang reviewed (commented):
LGTM overall, Vyas, it's in very good shape now. My comments below concerns mostly about wording only. 

Since I can't write diff for any parts of the text, let me leave notes here:
- line 122: we should mention `cuda-version` instead of `cudatoolkit`, now that CF has CUDA 12 support.

vyasr commented on docs/key-issues/gpus.md:
```diff
@@ -30,54 +30,78 @@ Packaging such projects for PyPI has been, and still is, quite challenging.
+supporting only a single CUDA version, and PyTorch providing more wheels for
+other CUDA versions and a CPU-only version in a separate wheelhouse (see
 [pytorch.org/get-started](https://pytorch.org/get-started/locally/)).
 CuPy provides a number of packages: [`cupy`](https://pypi.org/project/cupy/),
```
Fair point, avoiding mention of it does make sense.

vyasr commented:
> LGTM overall, Vyas, it's in very good shape now. My comments below concerns mostly about wording only.
> 
> Since I can't write diff for any parts of the text, let me leave notes here:
> 
>     * line 122: we should mention `cuda-version` instead of `cudatoolkit`, now that CF has CUDA 12 support.

Thanks Leo! I've addressed your comments. Also added discussion cuda-version. I didn't remove cudatookit altogether since I think many people will still be using CUDA 11, but I think we can consider removing it at a time when enough of the ecosystem has migrated to CUDA 12. Let me know what you think of that approach.

leofang reviewed (approved):
Thanks Vyas and all for help!

leofang commented:
@rgommers We think this is ready 🙂 PTAL, thx!

vyasr commented:
@h-vetinari if you'd like to take a final look as well please do!

leofang commented:
@rgommers friendly nudge 🙂

rgommers reviewed (approved):
Thank you @vyasr, @leofang, @gmarkall and @h-vetinari, this is great! Apologies for the long delay in getting this over the finish line. In it goes now - and I'll go point a few folks at this, the added detailed info on CUDA compatibility will be quite useful to folks working on GPU libraries I'm sure.

vyasr commented:
Thanks @rgommers!
